src/main/java/ws02/Vector.kt

```diff
@@ -0,0 +1,121 @@
+package ws02
+
+import java.lang.IllegalArgumentException
+import kotlin.math.pow
+import kotlin.math.sqrt
+
+open class Vector(initialSize : Int) : Comparable<Vector> {
+
+    private val _dimensions : MutableList<Double> = MutableList<Double>(initialSize) { _ -> 0.0 }
+
+    constructor(vararg values : Double) : this(values.size) {
+        var index = 0
+        for(value in values) {
+            _dimensions[index] = value
+            index++
+        }
+    }
+
+    val size : Int
+        get() = _dimensions.size
+
+    val length : Double
+        get() = sqrt(_dimensions.sumByDouble { value -> value.pow(2) })
+
+
+    fun getValues() : List<Double> = _dimensions.toList()
+
+    fun getValueOfDimension(dimensionIndex : Int) : Double {
+        var normalizedIndex = getNormalizedIndex(dimensionIndex)
+        ensureIndexInAvailableRange(normalizedIndex)
+        return getValueOfDimensionInternal(normalizedIndex)
+    }
+
+    private fun getValueOfDimensionInternal(index : Int) : Double {
+        return _dimensions[index]
+    }
+
+    fun setValueOfDimension(dimensionIndex: Int, value : Double) {
+        var normalizedIndex = getNormalizedIndex(dimensionIndex)
+        ensureIndexInAvailableRange(normalizedIndex)
+        setValueOfDimensionInternal(normalizedIndex, value)
+    }
+
+    private fun setValueOfDimensionInternal(index: Int, value : Double) {
+        _dimensions[index] = value
+    }
+
+    fun dotProductWith(anotherVector: Vector) : Double {
+        ensureSameSizeWith(anotherVector)
+        var product = 0.0
+        for(index in 0 until size) {
```

Comment:
Предлагаю использовать функции расширения для коллекций, чтобы перейти от императивного к декларативного стилю программирования:)
```suggestion
fun dotProductWith(anotherVector: Vector) : Double = _dimensions.mapIndexed { index, num -> 
        num * vector.dimensions[index] 
    }.fold(0.0) { acc: Double, num: Double ->
                acc + num
       }
           
```

Reply:
А как в декларативном стиле можно реализовать предварительную проверку размерности векторов?
Сделать свой экстеншн метод, что-то вроде .throwIf( condition ), который или возвращает исходную коллекцию, или кидает эксепшн?

Reply:
Когда, я делал замечание, то не пропустил момент проверки.
Сделать можно вот так. Но читаемость имхо упадет

`
    operator fun minus(anotherVector: Vector): Vector = takeIf { this.components.size == anotherVector.components.size }
        ?.components
        ?.mapIndexed { index, num ->
            num - anotherVector[index]
        }
        ?.asVector() ?: throw IllegalArgumentException("A vector's size should be the same")
`